enhancements/machine-api/external-remediations.md

```diff
@@ -0,0 +1,260 @@
+  - https://github.com/kubernetes-sigs/cluster-api/pull/3606
+---
+
+# External remediations
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+By using `MachineHealthChecks` a cluster admin can configure automatic remediation of unhealthy machines and nodes.
+The machine healthcheck controller's remediation strategy is deleting the machine, and letting the cloud provider
+create a new one. This isn't the best remediation strategy in all environments.
+
+There is already a mechanism to provide an alternative, external remediation strategy, by adding an annotation to the
+`MachineHealthCheck` and then to `Machine`s. However, this is isn't very maintainable and diverges from upstream.
+
+With this enhancement we propose a better, future-proof mechanism, that aligns us with the mechanism implemented upstream.
+This proposal is a backport of parts of the upstream machine healthcheck proposal [0], which
+also is already implemented [1].
+
+- [0] https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20191030-machine-health-checking.md
+- [1] https://github.com/kubernetes-sigs/cluster-api/pull/3606
+
+## Motivation
+
+- Environments consisting of hardware based clusters are significantly slower to (re)provision unhealthy machines,
+so they have a need for a remediation flow that includes at least one attempt at power-cycling unhealthy nodes.
+- Other environments and vendors also have specific remediation requirements, so there is a need to provide a generic
+mechanism for implementing custom remediation logic.
+
+### Goals
+
+- Create the ability to define customized remediation flows outside of the Machine Health Check and Machine API
+Operator codebase.
+- Migrate the existing external remediation mechanism to the new one.
+- Ensure that at most one external mechanism will attempt to recover the node.
+
+### Non-Goals
+
+TBD
+
+## Proposal
+
+We propose modifying the MachineHealthCheck CRD to support a externalRemediationTemplate, an ObjectReference to
+a provider-specific remediation template CR.
+
+### User Stories 
+
+#### Story 1
+
+As an admin of a hardware based cluster, I would like unhealthy nodes to be power-cycled, so that I can recover
+from transient errors faster and begin application recovery sooner.
+
+#### Story 2
+
+As an admin of a hardware based cluster, I would like unhealthy nodes to be power-cycled, so that I can detect
+non-transient issues faster.
+
+#### Story 3
+
+As an admin of a hardware based cluster, I would like the system to keep attempting to power-cycle unhealthy nodes,
+so that they are automatically added back to the cluster when I fix the underlying problem.
```

Comment:
Does attempting to power cycle while you are remediating the issue not actually make this problem worse? This sounds undesirable to me, if I'm working trying to fix a hardware issue, I don't want the machine to magically come back on mid way through the hardware change. 

Perhaps this story can be clarified a bit., I'm not huge on baremetal these days so I assume there's some nuances I'm not seeing here

Reply:
I believe the intention here is external issues, such as network problems (e.g. a host that can't reach the api-server).
TBH I'd expect the system to be able to recover itself even without power-cycle, so maybe this user story is not  very compelling

Reply:
I agree - removed